vault/expiration.go

```diff
@@ -114,6 +119,98 @@ func (c *Core) stopExpiration() error {
 	return nil
 }
 
+// Tidy cleans up stale leases which are associated with invalid tokens
+func (m *ExpirationManager) Tidy() error {
+	var tidyErrors *multierror.Error
+
+	// Create a cache to keep track of looked up tokens
+	tokenCache := make(map[string]bool)
+	i := 0
+
+	tidyFunc := func(leaseID string) {
+		i++
+		if i%500 == 0 {
+			m.logger.Debug("expiration: tidying of leases", "progress", i)
+		}
+
+		le, err := m.loadEntry(leaseID)
+		if err != nil {
+			tidyErrors = multierror.Append(tidyErrors, fmt.Errorf("failed to load the lease ID %q: %v", leaseID, err))
+			return
+		}
+
+		if le == nil {
+			tidyErrors = multierror.Append(tidyErrors, fmt.Errorf("nil entry for lease ID %q: %v", leaseID, err))
+			return
+		}
+
+		revokeLease := false
+		if le.ClientToken == "" {
+			m.logger.Debug("expiration: lease has an empty token", "lease_id", leaseID)
+			revokeLease = true
+		}
+
+		isValid, ok := tokenCache[le.ClientToken]
+		if !ok {
+			saltedID := m.tokenStore.SaltID(le.ClientToken)
+			lock := locksutil.LockForKey(m.tokenStore.tokenLocks, le.ClientToken)
+			lock.RLock()
+			te, err := m.tokenStore.lookupSalted(saltedID, true)
+			lock.RUnlock()
+
+			if err != nil {
+				tidyErrors = multierror.Append(tidyErrors, fmt.Errorf("failed to lookup token: %v", err))
+				return
+			}
+
+			if te == nil {
+				m.logger.Debug("expiration: lease has an invalid token", "lease_id", leaseID)
+				revokeLease = true
+				tokenCache[le.ClientToken] = false
+			} else {
+				tokenCache[le.ClientToken] = true
+			}
+		} else {
+			if isValid {
+				return
+			} else {
+				m.logger.Debug("expiration: lease has an invalid token", "lease_id", leaseID)
+				revokeLease = true
+			}
+		}
+
+		if revokeLease {
+			m.logger.Debug("expiration: invalid lease is being revoked", "lease_id", leaseID)
+			// Force the revocation and skip going through the token store
+			// again
+			err = m.revokeCommon(leaseID, true, true)
+			if err != nil {
+				tidyErrors = multierror.Append(tidyErrors, fmt.Errorf("failed to revoke an invalid lease with ID %q: %v", leaseID, err))
+				return
+			}
+		}
+	}
+
+	if atomic.CompareAndSwapInt64(&m.tidyLock, 0, 1) {
+		m.logger.Debug("expiration: beginning tidy operation on leases")
+		defer atomic.CompareAndSwapInt64(&m.tidyLock, 1, 0)
+		if err := logical.ScanView(m.idView, tidyFunc); err != nil {
+			return err
+		}
+	} else {
+		m.logger.Debug("expiration: tidy operation on leases is already in progress")
+		return fmt.Errorf("tidy operation on leases is already in progress")
+	}
+
+	// If no errors were encountered, return a normal error instead of a
+	// multierror
+	if tidyErrors == nil {
```

Comment:
Rather than this, just use https://godoc.org/github.com/hashicorp/go-multierror#Error.ErrorOrNil !